docs/3.1/docs/tsa-arch/arch-considerations-upgrading.rst

```diff
@@ -0,0 +1,265 @@
+* The ensure clause is also recomputed and re-evaluated for existing contracts when fetching or exercising choices on them.
+
+3.1.7 Adding Interface Definitions and Instances
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+* Interface definitions cannot be changed once deployed. Always place interface definitions in a standalone package containing only interfaces and no templates.
+* New interface instances can be added, but existing instances cannot be removed from a template. Interface choices can be made inoperable by having them evaluate to `error “No longer implemented.”`
+
+3.1.8 Adding and Deprecating Templates
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+New templates can be added, while existing templates cannot be removed but can be deprecated by:
+
+* Removing references to them from other Daml code.
+* Adding `ensure False` to make them non-operational. This prevents new contract creation using the template and choice exercises, including the implicit `Archive` choice, on existing contracts created using the template. 
+
+Note that the latter approach may result in a large number of active contracts stored on the ledger without a way to archive them, unless another update is deployed to evaluate the `ensure` clause to `True`. To deprecate a template without leaving contracts on the ledger that cannot be archived, add `ensure False` to the template only after all active contracts created from it have been archived through automation or other means.
+
+3.1.8.1 Understanding Contract Archival in Daml
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+At a purely technical level, "archived" means a consuming choice has been exercised on the contract. However, there are meaningful distinctions between the various business and operational contexts in which an archive may occur.
+
+1. Natural End of Lifecycle: The contract represents a business entity whose lifecycle has naturally ended. For example, a loan contract might be archived when the loan is fully repaid.
+
+2. State No Longer Holds True: The contract attested to a certain business state, but that state is no longer valid. For instance, an agreement that was conditional on certain criteria might be archived when those criteria are no longer met.
+
+3. Modification of the Underlying Entity/State: The business entity or state represented by the contract itself is still relevant; however, because Daml contracts are immutable, the update requires archiving the outdated contract. If the updated contract is written using v2, this results in an organic and incremental migration away from v1.
+
+4. Explicit Upgrade: The contract is archived as part of an upgrade process, preferably by an upgrade runner. This can be done during planned downtime, but in most cases, it can be done incrementally via a throttled background process. As a business operation, this is distinct from the previous three.
```

Comment:
Is it worth unpacking the why of "preferably by an upgrade runner" briefly here?